{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "This notebook demonstrates basic usage of BioThings Explorer, an engine for autonomously querying a distributed knowledge graph. BioThings Explorer can answer two classes of queries -- \"PREDICT\" and \"EXPLAIN\".  PREDICT queries are described in [PREDICT_demo.ipynb](PREDICT_demo.ipynb). Here, we describe EXPLAIN queries and how to use BioThings Explorer to execute them.  A more detailed overview of the BioThings Explorer systems is provided in [these slides](https://docs.google.com/presentation/d/1QWQqqQhPD_pzKryh6Wijm4YQswv8pAjleVORCPyJyDE/edit?usp=sharing).\n",
    "\n",
    "EXPLAIN queries are designed to **identify plausible reasoning chains to explain the relationship between two entities**.  For example, in this notebook, we explore the question:\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\"*Why does imatinib have an effect on the treatment of chronic myelogenous leukemia (CML)?*\"  \n",
    "\n",
    "Later, we also compare those results to a similar query looking at imatinib's role in treating gastrointestinal stromal tumors (GIST).\n",
    "\n",
    "**To experiment with an executable version of this notebook, [load it in Google Colaboratory](https://colab.research.google.com/github/biothings/biothings_explorer/blob/master/jupyter%20notebooks/EXPLAIN_demo.ipynb).**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 0: Load BioThings Explorer modules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, install the `biothings_explorer` and `biothings_schema` packages, as described in this [README](https://github.com/biothings/biothings_explorer/blob/master/jupyter%20notebooks/README.md#prerequisite).  This only needs to be done once (but including it here for compability with [colab](https://colab.research.google.com/))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "!pip install git+https://github.com/biothings/biothings_explorer#egg=biothings_explorer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, import the relevant modules:\n",
    "\n",
    "* **Hint**: Find corresponding bio-entity representation used in BioThings Explorer based on user input (could be any database IDs, symbols, names)\n",
    "* **FindConnection**: Find intermediate bio-entities which connects user specified input and output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import modules from biothings_explorer\n",
    "from biothings_explorer.hint import Hint\n",
    "from biothings_explorer.user_query_dispatcher import FindConnection\n",
    "import nest_asyncio\n",
    "nest_asyncio.apply()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1: Find representation of \"chronic myelogenous leukemia\" and \"imatinib\" in BTE\n",
    "\n",
    "In this step, BioThings Explorer translates our query strings \"chronic myelogenous leukemia\" and \"imatinib\" into BioThings objects, which contain mappings to many common identifiers.  Generally, the top result returned by the `Hint` module will be the correct item, but you should confirm that using the identifiers shown.\n",
    "\n",
    "Search terms can correspond to any child of [BiologicalEntity](https://biolink.github.io/biolink-model/docs/BiologicalEntity.html) from the [Biolink Model](https://biolink.github.io/biolink-model/docs/), including `DiseaseOrPhenotypicFeature` (e.g., \"lupus\"), `ChemicalSubstance` (e.g., \"acetaminophen\"), `Gene` (e.g., \"CDK2\"), `BiologicalProcess` (e.g., \"T cell differentiation\"), and `Pathway` (e.g., \"Citric acid cycle\")."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'MONDO': 'MONDO:0011996',\n",
       " 'DOID': 'DOID:8552',\n",
       " 'UMLS': 'C1292772',\n",
       " 'name': 'chronic myelogenous leukemia',\n",
       " 'MESH': 'D015464',\n",
       " 'OMIM': '608232',\n",
       " 'ORPHANET': '521',\n",
       " 'primary': {'identifier': 'MONDO',\n",
       "  'cls': 'Disease',\n",
       "  'value': 'MONDO:0011996'},\n",
       " 'display': 'MONDO(MONDO:0011996) DOID(DOID:8552) OMIM(608232) ORPHANET(521) UMLS(C1292772) MESH(D015464) name(chronic myelogenous leukemia)',\n",
       " 'type': 'Disease'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ht = Hint()\n",
    "# find all potential representations of CML\n",
    "cml_hint = ht.query(\"chronic myelogenous leukemia\")\n",
    "# select the correct representation of CML\n",
    "cml = cml_hint['Disease'][0]\n",
    "cml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'DRUGBANK': 'DB00619',\n",
       " 'CHEBI': 'CHEBI:45783',\n",
       " 'name': 'imatinib',\n",
       " 'primary': {'identifier': 'CHEBI',\n",
       "  'cls': 'ChemicalSubstance',\n",
       "  'value': 'CHEBI:45783'},\n",
       " 'display': 'CHEBI(CHEBI:45783) DRUGBANK(DB00619) name(imatinib)',\n",
       " 'type': 'ChemicalSubstance'}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find all potential representations of imatinib\n",
    "imatinib_hint = ht.query(\"imatinib\")\n",
    "# select the correct representation of imatinib\n",
    "imatinib = imatinib_hint['ChemicalSubstance'][0]\n",
    "imatinib"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2: Find intermediate nodes connecting imatinib and chronic myelogenous leukemia\n",
    "\n",
    "In this section, we find all paths in the knowledge graph that connect imatinib and chronic myelogenous leukemia.  To do that, we will use `FindConnection`.  This class is a convenient wrapper around two advanced functions for **query path planning** and **query path execution**. More advanced features for both query path planning and query path execution are in development and will be documented in the coming months. \n",
    "\n",
    "The parameters for `FindConnection` are described below:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function __init__ in module biothings_explorer.user_query_dispatcher:\n",
      "\n",
      "__init__(self, input_obj, output_obj, intermediate_nodes, registry=None)\n",
      "    Find relationships in the Knowledge Graph between an Input Object and an Output Object.\n",
      "    \n",
      "    Args:\n",
      "        input_obj (required): must be an object returned from Hint corresponding to a specific biomedical entity.\n",
      "                            Examples:\n",
      "                Hint().query(\"Fanconi anemia\")['DiseaseOrPhenotypicFeature'][0]\n",
      "                Hint().query(\"acetaminophen\")['ChemicalSubstance'][0]\n",
      "    \n",
      "        output_obj (required): must EITHER be an object returned from Hint corresponding to a specific biomedical\n",
      "                            entity, OR be a string or list of strings corresponding to Biolink Entity classes.\n",
      "                            Examples:\n",
      "                Hint().query(\"acetaminophen\")['ChemicalSubstance'][0]\n",
      "                'Gene'\n",
      "                ['Gene','ChemicalSubstance']\n",
      "    \n",
      "        intermediate_nodes (required): the semantic type(s) of the intermediate node(s).  Examples:\n",
      "                None                         : no intermediate node, find direct connections only\n",
      "                []                           : no intermediate node, find direct connections only\n",
      "                ['BiologicalEntity']         : one intermediate node of any semantic type\n",
      "                ['Gene']                     : one intermediate node that must be a Gene\n",
      "                [('Gene','Pathway')]         : one intermediate node that must be a Gene or a Pathway\n",
      "                ['Gene','Pathway']           : two intermediate nodes, first must be a Gene, second must be a Pathway.\n",
      "                ['Gene',('Pathway','Gene')]  : two intermediate nodes, first must be a Gene, second must be a Pathway or Gene.\n",
      "                                                **NOTE**: queries with more than one intermediate node are currently not supported\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(FindConnection.__init__)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we formulate a `FindConnection` query with \"CML\" as the `input_ojb`, \"imatinib\" as the `output_obj`.  We further specify with the `intermediate_nodes` parameter that we are looking for paths joining chronic myelogenous leukemia and imatinib with *one* intermediate node that is a Gene.  (The ability to search for longer reasoning paths that include additional intermediate nodes will be added shortly.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "fc = FindConnection(input_obj=cml, output_obj=imatinib, intermediate_nodes='Gene')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We next execute the `connect` method, which performs the **query path planning** and **query path execution** process.  In short, BioThings Explorer is deconstructing the query into individual API calls, executing those API calls, then assembling the results.\n",
    "\n",
    "A verbose log of this process is displayed below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==========\n",
      "========== QUERY PARAMETER SUMMARY ==========\n",
      "==========\n",
      "\n",
      "BTE will find paths that join 'chronic myelogenous leukemia' and 'imatinib'. Paths will have 1 intermediate node.\n",
      "\n",
      "Intermediate node #1 will have these type constraints: Gene\n",
      "\n",
      "\n",
      "==========\n",
      "========== QUERY #1 -- fetch all Gene entities linked to 'chronic myelogenous leukemia' ==========\n",
      "==========\n",
      "\n",
      "==== Step #1: Query path planning ====\n",
      "\n",
      "Because chronic myelogenous leukemia is of type 'Disease', BTE will query our meta-KG for APIs that can take 'Disease' as input and 'Gene' as output\n",
      "\n",
      "BTE found 10 apis:\n",
      "\n",
      "API 1. semmed_disease(15 API calls)\n",
      "API 2. cord_disease(1 API call)\n",
      "API 3. mgi_gene2phenotype(1 API call)\n",
      "API 4. scibite(1 API call)\n",
      "API 5. hetio(1 API call)\n",
      "API 6. mydisease(1 API call)\n",
      "API 7. DISEASES(1 API call)\n",
      "API 8. pharos(1 API call)\n",
      "API 9. scigraph(1 API call)\n",
      "API 10. biolink(1 API call)\n",
      "\n",
      "\n",
      "==== Step #2: Query path execution ====\n",
      "NOTE: API requests are dispatched in parallel, so the list of APIs below is ordered by query time.\n",
      "\n",
      "API 6.1: http://mydisease.info/v1/query?fields=disgenet.genes_related_to_disease.gene_id (POST -d q=C1292772,C0023473&scopes=mondo.xrefs.umls, disgenet.xrefs.umls)\n",
      "API 2.1: https://biothings.ncats.io/cord_disease/query?fields=associated_with (POST -d q=DOID:8552&scopes=doid)\n",
      "API 1.9: https://biothings.ncats.io/semmed/query?fields=disrupted_by (POST -d q=C1292772,C0023473&scopes=umls)\n",
      "API 1.11: https://biothings.ncats.io/semmed/query?fields=negatively_regulates (POST -d q=C1292772,C0023473&scopes=umls)\n",
      "API 1.5: https://biothings.ncats.io/semmed/query?fields=physically_interacts_with (POST -d q=C1292772,C0023473&scopes=umls)\n",
      "API 1.3: https://biothings.ncats.io/semmed/query?fields=prevented_by (POST -d q=C1292772,C0023473&scopes=umls)\n",
      "API 1.8: https://biothings.ncats.io/semmed/query?fields=derives_from (POST -d q=C1292772,C0023473&scopes=umls)\n",
      "API 1.4: https://biothings.ncats.io/semmed/query?fields=positively_regulates (POST -d q=C1292772,C0023473&scopes=umls)\n",
      "API 1.7: https://biothings.ncats.io/semmed/query?fields=positively_regulated_by (POST -d q=C1292772,C0023473&scopes=umls)\n",
      "API 1.13: https://biothings.ncats.io/semmed/query?fields=disrupts (POST -d q=C1292772,C0023473&scopes=umls)\n",
      "API 1.15: https://biothings.ncats.io/semmed/query?fields=negatively_regulated_by (POST -d q=C1292772,C0023473&scopes=umls)\n",
      "API 10.1: https://api.monarchinitiative.org/api/bioentity/disease/MONDO:0011996/genes?rows=200\n",
      "API 1.1: https://biothings.ncats.io/semmed/query?fields=affects (POST -d q=C1292772,C0023473&scopes=umls)\n",
      "API 1.14: https://biothings.ncats.io/semmed/query?fields=caused_by (POST -d q=C1292772,C0023473&scopes=umls)\n",
      "API 1.10: https://biothings.ncats.io/semmed/query?fields=affected_by (POST -d q=C1292772,C0023473&scopes=umls)\n",
      "API 7.1: https://pending.biothings.io/DISEASES/query?fields=DISEASES.associatedWith (POST -d q=DOID:8552&scopes=DISEASES.doid)\n",
      "API 3.1: https://pending.biothings.io/mgigene2phenotype/query?fields=_id&size=300 (POST -d q=DOID:8552&scopes=mgi.associated_with_disease.doid)\n",
      "API 1.2: https://biothings.ncats.io/semmed/query?fields=coexists_with (POST -d q=C1292772,C0023473&scopes=umls)\n",
      "API 1.6: https://biothings.ncats.io/semmed/query?fields=treated_by (POST -d q=C1292772,C0023473&scopes=umls)\n",
      "API 1.12: https://biothings.ncats.io/semmed/query?fields=related_to (POST -d q=C1292772,C0023473&scopes=umls)\n",
      "API 8.1: https://automat.renci.org/pharos/disease/gene/MONDO:0011996\n",
      "API 5.1: https://automat.renci.org/hetio/disease/gene/MONDO:0011996\n",
      "API 9.1: https://automat.renci.org/cord19_scigraph_v2/disease/gene/MONDO:0011996\n",
      "API 4.1: https://automat.renci.org/cord19_scibite_v2/disease/gene/MONDO:0011996\n",
      "\n",
      "\n",
      "==== Step #3: Output normalization ====\n",
      "\n",
      "API 3.1 mgi_gene2phenotype: 8 hits\n",
      "API 1.1 semmed_disease: No hits\n",
      "API 1.2 semmed_disease: No hits\n",
      "API 7.1 DISEASES: 3 hits\n",
      "API 1.3 semmed_disease: 7 hits\n",
      "API 1.4 semmed_disease: No hits\n",
      "API 1.5 semmed_disease: No hits\n",
      "API 1.6 semmed_disease: 133 hits\n",
      "API 1.7 semmed_disease: No hits\n",
      "API 8.1 pharos: 7 hits\n",
      "API 1.8 semmed_disease: No hits\n",
      "API 9.1 scigraph: 3 hits\n",
      "API 1.9 semmed_disease: 26 hits\n",
      "API 6.1 mydisease: 52 hits\n",
      "API 2.1 cord_disease: 146 hits\n",
      "API 1.10 semmed_disease: 65 hits\n",
      "API 4.1 scibite: 12 hits\n",
      "API 5.1 hetio: No hits\n",
      "API 1.11 semmed_disease: No hits\n",
      "API 1.12 semmed_disease: 550 hits\n",
      "API 1.13 semmed_disease: No hits\n",
      "API 1.14 semmed_disease: 90 hits\n",
      "API 10.1 biolink: 4 hits\n",
      "API 1.15 semmed_disease: No hits\n",
      "\n",
      "After id-to-object translation, BTE retrieved 759 unique objects.\n",
      "\n",
      "\n",
      "==========\n",
      "========== QUERY #2 -- fetch all Gene entities linked to 'imatinib' ==========\n",
      "==========\n",
      "\n",
      "==== Step #1: Query path planning ====\n",
      "\n",
      "Because imatinib is of type 'ChemicalSubstance', BTE will query our meta-KG for APIs that can take 'ChemicalSubstance' as input and 'Gene' as output\n",
      "\n",
      "BTE found 10 apis:\n",
      "\n",
      "API 1. mychem(3 API calls)\n",
      "API 2. ctd(2 API calls)\n",
      "API 3. semmed_chemical(13 API calls)\n",
      "API 4. chembio(1 API call)\n",
      "API 5. scibite(1 API call)\n",
      "API 6. cord_chemical(1 API call)\n",
      "API 7. pharos(1 API call)\n",
      "API 8. scigraph(1 API call)\n",
      "API 9. hmdb(1 API call)\n",
      "API 10. dgidb(2 API calls)\n",
      "\n",
      "\n",
      "==== Step #2: Query path execution ====\n",
      "NOTE: API requests are dispatched in parallel, so the list of APIs below is ordered by query time.\n",
      "\n",
      "API 3.4: https://biothings.ncats.io/semmedchemical/query?fields=disrupted_by (POST -d q=C0935989&scopes=umls)\n",
      "API 3.12: https://biothings.ncats.io/semmedchemical/query?fields=affected_by (POST -d q=C0935989&scopes=umls)\n",
      "API 2.2: http://ctdbase.org/tools/batchQuery.go?inputType=chem&report=genes_curated&format=json&inputTerms=C097613\n",
      "API 3.5: https://biothings.ncats.io/semmedchemical/query?fields=related_to (POST -d q=C0935989&scopes=umls)\n",
      "API 3.2: https://biothings.ncats.io/semmedchemical/query?fields=positively_regulates (POST -d q=C0935989&scopes=umls)\n",
      "API 3.13: https://biothings.ncats.io/semmedchemical/query?fields=positively_regulated_by (POST -d q=C0935989&scopes=umls)\n",
      "API 3.8: https://biothings.ncats.io/semmedchemical/query?fields=produces (POST -d q=C0935989&scopes=umls)\n",
      "API 3.9: https://biothings.ncats.io/semmedchemical/query?fields=produced_by (POST -d q=C0935989&scopes=umls)\n",
      "API 3.10: https://biothings.ncats.io/semmedchemical/query?fields=negatively_regulates (POST -d q=C0935989&scopes=umls)\n",
      "API 3.3: https://biothings.ncats.io/semmedchemical/query?fields=negatively_regulated_by (POST -d q=C0935989&scopes=umls)\n",
      "API 6.1: https://biothings.ncats.io/cord_chemical/query?fields=associated_with (POST -d q=CHEBI:45783&scopes=chebi)\n",
      "API 1.2: https://mychem.info/v1/query?fields=drugcentral.bioactivity (POST -d q=CHEMBL941,CHEMBL1642&scopes=chembl.molecule_chembl_id)\n",
      "API 2.1: http://ctdbase.org/tools/batchQuery.go?inputType=chem&report=genes_curated&format=json&inputTerms=D000068877\n",
      "API 1.1: https://mychem.info/v1/query?fields=drugbank.targets (POST -d q=DB00619&scopes=drugbank.id)\n",
      "API 1.3: https://mychem.info/v1/query?fields=drugbank.enzymes (POST -d q=DB00619&scopes=drugbank.id)\n",
      "API 3.7: https://biothings.ncats.io/semmedchemical/query?fields=coexists_with (POST -d q=C0935989&scopes=umls)\n",
      "API 3.11: https://biothings.ncats.io/semmedchemical/query?fields=affects (POST -d q=C0935989&scopes=umls)\n",
      "API 10.1: http://dgidb.genome.wustl.edu/api/v2/interactions.json?drugs=CHEMBL941\n",
      "API 3.6: https://biothings.ncats.io/semmedchemical/query?fields=disrupts (POST -d q=C0935989&scopes=umls)\n",
      "API 10.2: http://dgidb.genome.wustl.edu/api/v2/interactions.json?drugs=CHEMBL1642\n",
      "API 3.1: https://biothings.ncats.io/semmedchemical/query?fields=physically_interacts_with (POST -d q=C0935989&scopes=umls)\n",
      "API 8.1: https://automat.renci.org/cord19_scigraph_v2/chemical_substance/gene/CHEBI:45783\n",
      "API 5.1: https://automat.renci.org/cord19_scibite_v2/chemical_substance/gene/CHEBI:45783\n",
      "API 4.1: https://automat.renci.org/chembio/chemical_substance/gene/CHEBI:45783\n",
      "API 9.1: https://automat.renci.org/hmdb/chemical_substance/gene/CHEBI:45783\n",
      "API 7.1: https://automat.renci.org/pharos/chemical_substance/gene/CHEBI:45783\n",
      "\n",
      "\n",
      "==== Step #3: Output normalization ====\n",
      "\n",
      "API 3.1 semmed_chemical: 300 hits\n",
      "API 4.1 chembio: No hits\n",
      "API 1.1 mychem: 9 hits\n",
      "API 1.2 mychem: 80 hits\n",
      "API 1.3 mychem: 9 hits\n",
      "API 5.1 scibite: 8 hits\n",
      "API 7.1 pharos: 6 hits\n",
      "API 3.2 semmed_chemical: 98 hits\n",
      "API 3.3 semmed_chemical: 38 hits\n",
      "API 3.4 semmed_chemical: No hits\n",
      "API 3.5 semmed_chemical: No hits\n",
      "API 8.1 scigraph: 3 hits\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "API 10.1 dgidb: 34 hits\n",
      "API 10.2 dgidb: 5 hits\n",
      "API 9.1 hmdb: No hits\n",
      "API 3.6 semmed_chemical: No hits\n",
      "API 3.7 semmed_chemical: 138 hits\n",
      "API 3.8 semmed_chemical: No hits\n",
      "API 3.9 semmed_chemical: 12 hits\n",
      "API 6.1 cord_chemical: 172 hits\n",
      "API 2.1 ctd: 199 hits\n",
      "API 2.2 ctd: No hits\n",
      "API 3.10 semmed_chemical: 171 hits\n",
      "API 3.11 semmed_chemical: No hits\n",
      "API 3.12 semmed_chemical: No hits\n",
      "API 3.13 semmed_chemical: 39 hits\n",
      "\n",
      "After id-to-object translation, BTE retrieved 730 unique objects.\n",
      "\n",
      "==========\n",
      "========== Final assembly of results ==========\n",
      "==========\n",
      "\n",
      "\n",
      "BTE found 255 unique intermediate nodes connecting 'chronic myelogenous leukemia' and 'imatinib'\n"
     ]
    }
   ],
   "source": [
    "# set verbose=True will display all steps which BTE takes to find the connection\n",
    "fc.connect(verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3: Display and Filter results\n",
    "This section demonstrates post-query filtering done in Python. Later, more advanced filtering functions will be added to the **query path execution** module for interleaved filtering, thereby enabling longer query paths. More details to come...\n",
    "\n",
    "First, all matching paths can be exported to a data frame. Let's examine a sample of those results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1621, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input</th>\n",
       "      <th>input_type</th>\n",
       "      <th>pred1</th>\n",
       "      <th>pred1_source</th>\n",
       "      <th>pred1_api</th>\n",
       "      <th>pred1_pubmed</th>\n",
       "      <th>node1_type</th>\n",
       "      <th>node1_name</th>\n",
       "      <th>node1_id</th>\n",
       "      <th>pred2</th>\n",
       "      <th>pred2_source</th>\n",
       "      <th>pred2_api</th>\n",
       "      <th>pred2_pubmed</th>\n",
       "      <th>output_type</th>\n",
       "      <th>output_name</th>\n",
       "      <th>output_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1125</td>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>18406870</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>NCBIGene:25</td>\n",
       "      <td>positively_regulated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>26251899</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>725</td>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>BioLink API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BCR</td>\n",
       "      <td>NCBIGene:613</td>\n",
       "      <td>negatively_regulated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>19307018</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>816</td>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>pharos</td>\n",
       "      <td>Automat PHAROS API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BCR</td>\n",
       "      <td>NCBIGene:613</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>DGIdb API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1537</td>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>1815390,11587372,10676660,10092207,11905636,12...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>NCBIGene:4547</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>12669727,15283151,21505592,23434731,22713161</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>803</td>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>scibite</td>\n",
       "      <td>Automat CORD19 Scibite API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BCR</td>\n",
       "      <td>NCBIGene:613</td>\n",
       "      <td>positively_regulated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>16343892,22904675,24269846,27444277</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1695</td>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>23818300,26966074</td>\n",
       "      <td>Gene</td>\n",
       "      <td>TP53</td>\n",
       "      <td>NCBIGene:7157</td>\n",
       "      <td>positively_regulates</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>20094798,23598363,25280212</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1532</td>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>1815390,11587372,10676660,10092207,11905636,12...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>NCBIGene:4547</td>\n",
       "      <td>negatively_regulated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>23434731</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>589</td>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>8161775,2395384,12829610,16045749,20520635</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BCR</td>\n",
       "      <td>NCBIGene:613</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>drugcentral</td>\n",
       "      <td>MyChem.info API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1890</td>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Translator Text Mining Provider</td>\n",
       "      <td>CORD Disease API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ESR1</td>\n",
       "      <td>NCBIGene:2099</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>12783377,12682876,24860788</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1417</td>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Translator Text Mining Provider</td>\n",
       "      <td>CORD Disease API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>7322</td>\n",
       "      <td>HGNC:7322</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Translator Text Mining Provider</td>\n",
       "      <td>CORD Chemical API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             input input_type        pred1  \\\n",
       "1125  CHRONIC MYELOGENOUS LEUKEMIA    Disease   related_to   \n",
       "725   CHRONIC MYELOGENOUS LEUKEMIA    Disease   related_to   \n",
       "816   CHRONIC MYELOGENOUS LEUKEMIA    Disease   related_to   \n",
       "1537  CHRONIC MYELOGENOUS LEUKEMIA    Disease    caused_by   \n",
       "803   CHRONIC MYELOGENOUS LEUKEMIA    Disease   related_to   \n",
       "1695  CHRONIC MYELOGENOUS LEUKEMIA    Disease   treated_by   \n",
       "1532  CHRONIC MYELOGENOUS LEUKEMIA    Disease    caused_by   \n",
       "589   CHRONIC MYELOGENOUS LEUKEMIA    Disease  affected_by   \n",
       "1890  CHRONIC MYELOGENOUS LEUKEMIA    Disease   related_to   \n",
       "1417  CHRONIC MYELOGENOUS LEUKEMIA    Disease   related_to   \n",
       "\n",
       "                         pred1_source                   pred1_api  \\\n",
       "1125                           SEMMED          SEMMED Disease API   \n",
       "725                              None                 BioLink API   \n",
       "816                            pharos          Automat PHAROS API   \n",
       "1537                           SEMMED          SEMMED Disease API   \n",
       "803                           scibite  Automat CORD19 Scibite API   \n",
       "1695                           SEMMED          SEMMED Disease API   \n",
       "1532                           SEMMED          SEMMED Disease API   \n",
       "589                            SEMMED          SEMMED Disease API   \n",
       "1890  Translator Text Mining Provider            CORD Disease API   \n",
       "1417  Translator Text Mining Provider            CORD Disease API   \n",
       "\n",
       "                                           pred1_pubmed node1_type node1_name  \\\n",
       "1125                                           18406870       Gene       ABL1   \n",
       "725                                                None       Gene        BCR   \n",
       "816                                                None       Gene        BCR   \n",
       "1537  1815390,11587372,10676660,10092207,11905636,12...       Gene       MTTP   \n",
       "803                                                None       Gene        BCR   \n",
       "1695                                  23818300,26966074       Gene       TP53   \n",
       "1532  1815390,11587372,10676660,10092207,11905636,12...       Gene       MTTP   \n",
       "589          8161775,2395384,12829610,16045749,20520635       Gene        BCR   \n",
       "1890                                               None       Gene       ESR1   \n",
       "1417                                               None       Gene       7322   \n",
       "\n",
       "           node1_id                      pred2  \\\n",
       "1125    NCBIGene:25    positively_regulated_by   \n",
       "725    NCBIGene:613    negatively_regulated_by   \n",
       "816    NCBIGene:613  physically_interacts_with   \n",
       "1537  NCBIGene:4547              coexists_with   \n",
       "803    NCBIGene:613    positively_regulated_by   \n",
       "1695  NCBIGene:7157       positively_regulates   \n",
       "1532  NCBIGene:4547    negatively_regulated_by   \n",
       "589    NCBIGene:613  physically_interacts_with   \n",
       "1890  NCBIGene:2099              coexists_with   \n",
       "1417      HGNC:7322                 related_to   \n",
       "\n",
       "                         pred2_source            pred2_api  \\\n",
       "1125                           SEMMED  SEMMED Chemical API   \n",
       "725                            SEMMED  SEMMED Chemical API   \n",
       "816                              None            DGIdb API   \n",
       "1537                           SEMMED  SEMMED Chemical API   \n",
       "803                            SEMMED  SEMMED Chemical API   \n",
       "1695                           SEMMED  SEMMED Chemical API   \n",
       "1532                           SEMMED  SEMMED Chemical API   \n",
       "589                       drugcentral      MyChem.info API   \n",
       "1890                           SEMMED  SEMMED Chemical API   \n",
       "1417  Translator Text Mining Provider    CORD Chemical API   \n",
       "\n",
       "                                      pred2_pubmed output_type output_name  \\\n",
       "1125                                      26251899        Gene    IMATINIB   \n",
       "725                                       19307018        Gene    IMATINIB   \n",
       "816                                           None        Gene    IMATINIB   \n",
       "1537  12669727,15283151,21505592,23434731,22713161        Gene    IMATINIB   \n",
       "803            16343892,22904675,24269846,27444277        Gene    IMATINIB   \n",
       "1695                    20094798,23598363,25280212        Gene    IMATINIB   \n",
       "1532                                      23434731        Gene    IMATINIB   \n",
       "589                                           None        Gene    IMATINIB   \n",
       "1890                    12783377,12682876,24860788        Gene    IMATINIB   \n",
       "1417                                          None        Gene    IMATINIB   \n",
       "\n",
       "          output_id  \n",
       "1125  name:IMATINIB  \n",
       "725   name:IMATINIB  \n",
       "816   name:IMATINIB  \n",
       "1537  name:IMATINIB  \n",
       "803   name:IMATINIB  \n",
       "1695  name:IMATINIB  \n",
       "1532  name:IMATINIB  \n",
       "589   name:IMATINIB  \n",
       "1890  name:IMATINIB  \n",
       "1417  name:IMATINIB  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = fc.display_table_view()\n",
    "\n",
    "# because UMLS is not currently well-integrated in our ID-to-object translation system, removing UMLS-only entries here\n",
    "patternDel = \"^UMLS:C\\d+\"\n",
    "filter = df.node1_id.str.contains(patternDel)\n",
    "df = df[~filter]\n",
    "\n",
    "print(df.shape)\n",
    "df.sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While most results are based on edges from [semmed](https://skr3.nlm.nih.gov/SemMed/), edges from [DGIdb](http://www.dgidb.org/), [biolink](https://monarchinitiative.org/), [disgenet](http://www.disgenet.org/), [mydisease.info](https://mydisease.info) and [drugcentral](http://drugcentral.org/) were also retrieved from their respective APIs.  \n",
    "\n",
    "Next, let's look to see which genes are mentioned the most."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BCR      273\n",
       "KIT      135\n",
       "ABL1     126\n",
       "MTTP      45\n",
       "AKT1      44\n",
       "ABCB1     39\n",
       "TP53      35\n",
       "CD34      24\n",
       "ABCG2     24\n",
       "LYN       21\n",
       "Name: node1_name, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.node1_name.value_counts().head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not surprisingly, the top two genes that BioThings Explorer found that join imatinib to CML are *ABL1* and *BCR*, the two genes that are fused in the \"Philadelphia chromosome\", the genetic abnormality that underlies CML, *and* the validate target of imatinib.\n",
    "\n",
    "Let's examine some of the PubMed articles linking **CML to *ABL1*** and ***ABL1* to imatinib**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 17 articles supporting the edge between CML and ABL1. Sampling of 10 of those:\n",
      "http://pubmed.gov/24662807\n",
      "http://pubmed.gov/26179066\n",
      "http://pubmed.gov/11979553\n",
      "http://pubmed.gov/10498618\n",
      "http://pubmed.gov/10822991\n",
      "http://pubmed.gov/11368359\n",
      "http://pubmed.gov/20809971\n",
      "http://pubmed.gov/18082628\n",
      "http://pubmed.gov/18243808\n",
      "http://pubmed.gov/23287430\n"
     ]
    }
   ],
   "source": [
    "# fetch all articles connecting 'chronic myelogenous leukemia' and 'ABL1'\n",
    "articles = []\n",
    "for info in fc.display_edge_info('chronic myelogenous leukemia', 'ABL1').values():\n",
    "    if 'pubmed' in info['info']:\n",
    "        articles += info['info']['pubmed']\n",
    "print(\"There are \"+str(len(articles))+\" articles supporting the edge between CML and ABL1. Sampling of 10 of those:\")\n",
    "x = [print(\"http://pubmed.gov/\"+str(x)) for x in articles[0:10] ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 32 articles supporting the edge between ABL1 and imatinib. Sampling of 10 of those:\n",
      "http://pubmed.gov/15799618\n",
      "http://pubmed.gov/15917650\n",
      "http://pubmed.gov/15949566\n",
      "http://pubmed.gov/16153117\n",
      "http://pubmed.gov/16205964\n",
      "http://pubmed.gov/15713800\n",
      "http://pubmed.gov/15713800\n",
      "http://pubmed.gov/15713800\n",
      "http://pubmed.gov/19166098\n",
      "http://pubmed.gov/26030291\n"
     ]
    }
   ],
   "source": [
    "# fetch all articles connecting 'ABL1' and 'Imatinib\n",
    "articles = []\n",
    "for info in fc.display_edge_info('ABL1', 'imatinib').values():\n",
    "    if 'pubmed' in info['info']:\n",
    "        articles += info['info']['pubmed']\n",
    "print(\"There are \"+str(len(articles))+\" articles supporting the edge between ABL1 and imatinib. Sampling of 10 of those:\")\n",
    "x = [print(\"http://pubmed.gov/\"+str(x)) for x in articles[0:10] ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparing results between CML and GIST\n",
    "\n",
    "Let's perform another BioThings Explorer query, this time looking to EXPLAIN the relationship between imatinib and gastrointestinal stromal tumors (GIST), another disease treated by imatinib."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'MONDO': 'MONDO:0011719',\n",
       " 'DOID': 'DOID:9253',\n",
       " 'UMLS': 'C3179349',\n",
       " 'name': 'gastrointestinal stromal tumor',\n",
       " 'MESH': 'D046152',\n",
       " 'OMIM': '606764',\n",
       " 'ORPHANET': '44890',\n",
       " 'primary': {'identifier': 'MONDO',\n",
       "  'cls': 'Disease',\n",
       "  'value': 'MONDO:0011719'},\n",
       " 'display': 'MONDO(MONDO:0011719) DOID(DOID:9253) OMIM(606764) ORPHANET(44890) UMLS(C3179349) MESH(D046152) name(gastrointestinal stromal tumor)',\n",
       " 'type': 'Disease'}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ht = Hint()\n",
    "# find all potential representations of CML\n",
    "gist_hint = ht.query(\"gastrointestinal stromal tumor\")\n",
    "# select the correct representation of CML\n",
    "gist = gist_hint['Disease'][0]\n",
    "gist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "fc = FindConnection(input_obj=gist, output_obj=imatinib, intermediate_nodes='Gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "fc.connect(verbose=False) # skipping the verbose log here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1757, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input</th>\n",
       "      <th>input_type</th>\n",
       "      <th>pred1</th>\n",
       "      <th>pred1_source</th>\n",
       "      <th>pred1_api</th>\n",
       "      <th>pred1_pubmed</th>\n",
       "      <th>node1_type</th>\n",
       "      <th>node1_name</th>\n",
       "      <th>node1_id</th>\n",
       "      <th>pred2</th>\n",
       "      <th>pred2_source</th>\n",
       "      <th>pred2_api</th>\n",
       "      <th>pred2_pubmed</th>\n",
       "      <th>output_type</th>\n",
       "      <th>output_name</th>\n",
       "      <th>output_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1076</td>\n",
       "      <td>GASTROINTESTINAL STROMAL SARCOMA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>17193822,16900856,15838387,15685537,15917417,1...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>NCBIGene:5156</td>\n",
       "      <td>negatively_regulates</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>24963404,28020350,28768491,17437861</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>463</td>\n",
       "      <td>GASTROINTESTINAL STROMAL SARCOMA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>DISEASE</td>\n",
       "      <td>DISEASES API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>KIT</td>\n",
       "      <td>NCBIGene:3815</td>\n",
       "      <td>negatively_regulated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>20425130,20043176,20109338,26722383</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1030</td>\n",
       "      <td>GASTROINTESTINAL STROMAL SARCOMA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>BioLink API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>NCBIGene:5156</td>\n",
       "      <td>positively_regulates</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>21828142</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>715</td>\n",
       "      <td>GASTROINTESTINAL STROMAL SARCOMA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>21666577</td>\n",
       "      <td>Gene</td>\n",
       "      <td>KIT</td>\n",
       "      <td>NCBIGene:3815</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>17438095,20975605,23787115</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1247</td>\n",
       "      <td>GASTROINTESTINAL STROMAL SARCOMA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>15297464</td>\n",
       "      <td>Gene</td>\n",
       "      <td>FLT3</td>\n",
       "      <td>NCBIGene:2322</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>14976243</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>483</td>\n",
       "      <td>GASTROINTESTINAL STROMAL SARCOMA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>disgenet</td>\n",
       "      <td>mydisease.info API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>KIT</td>\n",
       "      <td>NCBIGene:3815</td>\n",
       "      <td>negatively_regulated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>20425130,20043176,20109338,26722383</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>617</td>\n",
       "      <td>GASTROINTESTINAL STROMAL SARCOMA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>12938260,10779223,10779223,11706520,27771813</td>\n",
       "      <td>Gene</td>\n",
       "      <td>KIT</td>\n",
       "      <td>NCBIGene:3815</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>12441322,15846297,16797704,21586300,21249321,2...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1372</td>\n",
       "      <td>GASTROINTESTINAL STROMAL SARCOMA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>20489620,26098203</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PDGFA</td>\n",
       "      <td>NCBIGene:5154</td>\n",
       "      <td>negatively_regulates</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>18312355</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>603</td>\n",
       "      <td>GASTROINTESTINAL STROMAL SARCOMA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>disgenet</td>\n",
       "      <td>mydisease.info API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>KIT</td>\n",
       "      <td>NCBIGene:3815</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>12481435,12969987,16087693,20109338,25985771</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>888</td>\n",
       "      <td>GASTROINTESTINAL STROMAL SARCOMA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>12672043,16077968,17582306,18486988,21300610</td>\n",
       "      <td>Gene</td>\n",
       "      <td>KIT</td>\n",
       "      <td>NCBIGene:3815</td>\n",
       "      <td>related_to</td>\n",
       "      <td>CTD</td>\n",
       "      <td>CTD API</td>\n",
       "      <td>21295132</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 input input_type       pred1 pred1_source  \\\n",
       "1076  GASTROINTESTINAL STROMAL SARCOMA    Disease  related_to       SEMMED   \n",
       "463   GASTROINTESTINAL STROMAL SARCOMA    Disease  related_to      DISEASE   \n",
       "1030  GASTROINTESTINAL STROMAL SARCOMA    Disease  related_to         None   \n",
       "715   GASTROINTESTINAL STROMAL SARCOMA    Disease  treated_by       SEMMED   \n",
       "1247  GASTROINTESTINAL STROMAL SARCOMA    Disease  related_to       SEMMED   \n",
       "483   GASTROINTESTINAL STROMAL SARCOMA    Disease  related_to     disgenet   \n",
       "617   GASTROINTESTINAL STROMAL SARCOMA    Disease   caused_by       SEMMED   \n",
       "1372  GASTROINTESTINAL STROMAL SARCOMA    Disease  treated_by       SEMMED   \n",
       "603   GASTROINTESTINAL STROMAL SARCOMA    Disease  related_to     disgenet   \n",
       "888   GASTROINTESTINAL STROMAL SARCOMA    Disease  related_to       SEMMED   \n",
       "\n",
       "               pred1_api                                       pred1_pubmed  \\\n",
       "1076  SEMMED Disease API  17193822,16900856,15838387,15685537,15917417,1...   \n",
       "463         DISEASES API                                               None   \n",
       "1030         BioLink API                                               None   \n",
       "715   SEMMED Disease API                                           21666577   \n",
       "1247  SEMMED Disease API                                           15297464   \n",
       "483   mydisease.info API                                               None   \n",
       "617   SEMMED Disease API       12938260,10779223,10779223,11706520,27771813   \n",
       "1372  SEMMED Disease API                                  20489620,26098203   \n",
       "603   mydisease.info API                                               None   \n",
       "888   SEMMED Disease API       12672043,16077968,17582306,18486988,21300610   \n",
       "\n",
       "     node1_type node1_name       node1_id                      pred2  \\\n",
       "1076       Gene     PDGFRA  NCBIGene:5156       negatively_regulates   \n",
       "463        Gene        KIT  NCBIGene:3815    negatively_regulated_by   \n",
       "1030       Gene     PDGFRA  NCBIGene:5156       positively_regulates   \n",
       "715        Gene        KIT  NCBIGene:3815              coexists_with   \n",
       "1247       Gene       FLT3  NCBIGene:2322  physically_interacts_with   \n",
       "483        Gene        KIT  NCBIGene:3815    negatively_regulated_by   \n",
       "617        Gene        KIT  NCBIGene:3815  physically_interacts_with   \n",
       "1372       Gene      PDGFA  NCBIGene:5154       negatively_regulates   \n",
       "603        Gene        KIT  NCBIGene:3815  physically_interacts_with   \n",
       "888        Gene        KIT  NCBIGene:3815                 related_to   \n",
       "\n",
       "     pred2_source            pred2_api  \\\n",
       "1076       SEMMED  SEMMED Chemical API   \n",
       "463        SEMMED  SEMMED Chemical API   \n",
       "1030       SEMMED  SEMMED Chemical API   \n",
       "715        SEMMED  SEMMED Chemical API   \n",
       "1247       SEMMED  SEMMED Chemical API   \n",
       "483        SEMMED  SEMMED Chemical API   \n",
       "617        SEMMED  SEMMED Chemical API   \n",
       "1372       SEMMED  SEMMED Chemical API   \n",
       "603        SEMMED  SEMMED Chemical API   \n",
       "888           CTD              CTD API   \n",
       "\n",
       "                                           pred2_pubmed output_type  \\\n",
       "1076                24963404,28020350,28768491,17437861        Gene   \n",
       "463                 20425130,20043176,20109338,26722383        Gene   \n",
       "1030                                           21828142        Gene   \n",
       "715                          17438095,20975605,23787115        Gene   \n",
       "1247                                           14976243        Gene   \n",
       "483                 20425130,20043176,20109338,26722383        Gene   \n",
       "617   12441322,15846297,16797704,21586300,21249321,2...        Gene   \n",
       "1372                                           18312355        Gene   \n",
       "603        12481435,12969987,16087693,20109338,25985771        Gene   \n",
       "888                                            21295132        Gene   \n",
       "\n",
       "     output_name      output_id  \n",
       "1076    IMATINIB  name:IMATINIB  \n",
       "463     IMATINIB  name:IMATINIB  \n",
       "1030    IMATINIB  name:IMATINIB  \n",
       "715     IMATINIB  name:IMATINIB  \n",
       "1247    IMATINIB  name:IMATINIB  \n",
       "483     IMATINIB  name:IMATINIB  \n",
       "617     IMATINIB  name:IMATINIB  \n",
       "1372    IMATINIB  name:IMATINIB  \n",
       "603     IMATINIB  name:IMATINIB  \n",
       "888     IMATINIB  name:IMATINIB  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = fc.display_table_view()\n",
    "\n",
    "# because UMLS is not currently well-integrated in our ID-to-object translation system, removing UMLS-only entries here\n",
    "patternDel = \"^UMLS:C\\d+\"\n",
    "filter = df.node1_id.str.contains(patternDel)\n",
    "df = df[~filter]\n",
    "\n",
    "print(df.shape)\n",
    "df.sample(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "KIT       594\n",
       "PDGFRA    192\n",
       "BCR        63\n",
       "ABL1       28\n",
       "TP53       28\n",
       "VEGFA      24\n",
       "CD34       24\n",
       "EGFR       18\n",
       "MTTP       18\n",
       "BRAF       16\n",
       "Name: node1_name, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.node1_name.value_counts().head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, the top two genes that BioThings Explorer found that join imatinib to GIST are *PDGFRA* and *KIT*, the most commonly mutated genes found in GIST *and* validated targets of imatinib.\n",
    "\n",
    "While several of the listed genes would be considered positive controls, others on the list could be viewed as **testable hypotheses and discovery opportunities** to be evaluated by domain experts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusions and caveats\n",
    "\n",
    "This notebook demonstrated the use of BioThings Explorer in EXPLAIN mode to investigate the relationship between imatinib and two diseases that it treats -- chronic myelogenous leukemia (CML) and gastrointestinal stromal tumors (GIST).  In each case, BioThings Explorer autonomously queried a **distributed knowledge graph of biomedical APIs** to find the most common genes, and in each case the relevant targets were retrieved.\n",
    "\n",
    "There are still many areas for improvement (and some areas in which BioThings Explorer is still buggy).  And of course, BioThings Explorer is dependent on the accessibility of the APIs that comprise the distributed knowledge graph.  Nevertheless, we encourage users to try other variants of the EXPLAIN queries demonstrated in this notebook."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
